File: android/app/src/main/java/com/memorize/learning/TextComparator.kt
package com.memorize.learning

object TextComparator {
    private val dontRememberPhrases = listOf(
        "не помню", "непомню", "не знаю", "незнаю",
        "забыл", "забыла", "не помню", "не могу вспомнить"
    )
    
    fun isDontRemember(text: String): Boolean {
        val normalized = text.lowercase().trim()
        return dontRememberPhrases.any { normalized.contains(it) }
    }
    
    fun compareTexts(original: String, userText: String): Boolean {
        val normalizedOriginal = normalizeText(original)
        val normalizedUser = normalizeText(userText)
        
        // Exact match
        if (normalizedOriginal == normalizedUser) return true
        
        // Calculate similarity using Levenshtein distance
        val similarity = calculateSimilarity(normalizedOriginal, normalizedUser)
        
        // Consider correct if similarity is above 85%
        return similarity >= 0.85
    }
    
    private fun normalizeText(text: String): String {
        return text.lowercase()
            .replace(Regex("[.,!?;:()\\[\\]{}'\"]"), "")
            .replace(Regex("\\s+"), " ")
            .trim()
    }
    
    private fun calculateSimilarity(s1: String, s2: String): Double {
        val maxLength = maxOf(s1.length, s2.length)
        if (maxLength == 0) return 1.0
        
        val distance = levenshteinDistance(s1, s2)
        return 1.0 - (distance.toDouble() / maxLength)
    }
    
    private fun levenshteinDistance(s1: String, s2: String): Int {
        val m = s1.length
        val n = s2.length
        val dp = Array(m + 1) { IntArray(n + 1) }
        
        for (i in 0..m) dp[i][0] = i
        for (j in 0..n) dp[0][j] = j
        
        for (i in 1..m) {
            for (j in 1..n) {
                val cost = if (s1[i - 1] == s2[j - 1]) 0 else 1
                dp[i][j] = minOf(
                    dp[i - 1][j] + 1,
                    dp[i][j - 1] + 1,
                    dp[i - 1][j - 1] + cost
                )
            }
        }
        
        return dp[m][n]
    }
}

File: ios/Memorize/Learning/TextComparator.swift
import Foundation

class TextComparator {
    private static let dontRememberPhrases = [
        "не помню", "непомню", "не знаю", "незнаю",
        "забыл", "забыла", "не могу вспомнить"
    ]
    
    static func isDontRemember(_ text: String) -> Bool {
        let normalized = text.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        return dontRememberPhrases.contains { normalized.contains($0) }
    }
    
    static func compareTexts(original: String, userText: String) -> Bool {
        let normalizedOriginal = normalizeText(original)
        let normalizedUser = normalizeText(userText)
        
        if normalizedOriginal == normalizedUser {
            return true
        }
        
        let similarity = calculateSimilarity(normalizedOriginal, normalizedUser)
        return similarity >= 0.85
    }
    
    private static func normalizeText(_ text: String) -> String {
        return text.lowercased()
            .replacingOccurrences(of: "[.,!?;:()\\[\\]{}'\"]", with: "", options: .regularExpression)
            .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
    
    private static func calculateSimilarity(_ s1: String, _ s2: String) -> Double {
        let maxLength = max(s1.count, s2.count)
        if maxLength == 0 { return 1.0 }
        
        let distance = levenshteinDistance(s1, s2)
        return 1.0 - (Double(distance) / Double(maxLength))
    }
    
    private static func levenshteinDistance(_ s1: String, _ s2: String) -> Int {
        let m = s1.count
        let n = s2.count
        var dp = Array(repeating: Array(repeating: 0, count: n + 1), count: m + 1)
        
        for i in 0...m { dp[i][0] = i }
        for j in 0...n { dp[0][j] = j }
        
        let s1Array = Array(s1)
        let s2Array = Array(s2)
        
        for i in 1...m {
            for j in 1...n {
                let cost = s1Array[i - 1] == s2Array[j - 1] ? 0 : 1
                dp[i][j] = min(
                    dp[i - 1][j] + 1,
                    dp[i][j - 1] + 1,
                    dp[i - 1][j - 1] + cost
                )
            }
        }
        
        return dp[m][n]
    }
}

